### PR TITLE
feat(starlette/fastapi): trace background tasks using span links

### PR DESCRIPTION
**Pull request by mabdinur**
Resolves https://github.com/DataDog/dd-trace-py/issues/7132 and https://github.com/DataDog/dd-trace-py/issues/3851 , by ensuring background tasks are not nested under fastapi/starlette request spans. Instead traced background tasks linked to request spans via span links. 

This PR introduces the first usage of span links in a ddtrace integration. H̶o̶w̶e̶v̶e̶r̶ ̶s̶p̶a̶n̶ ̶l̶i̶n̶k̶s̶ ̶a̶r̶e̶ ̶n̶o̶t̶ ̶y̶e̶t̶ ̶s̶u̶p̶p̶o̶r̶t̶e̶d̶ ̶b̶y̶ ̶t̶h̶e̶ ̶d̶a̶t̶a̶d̶o̶g̶ ̶a̶g̶e̶n̶t̶.̶ ̶S̶p̶a̶n̶ ̶l̶i̶n̶k̶s̶ ̶g̶e̶n̶e̶r̶a̶t̶e̶d̶ ̶b̶y̶ ̶f̶a̶s̶t̶a̶p̶i̶ ̶a̶n̶d̶ ̶s̶t̶a̶r̶l̶e̶t̶t̶e̶ ̶i̶n̶t̶e̶g̶r̶a̶t̶i̶o̶n̶s̶ ̶w̶i̶l̶l̶ ̶n̶o̶t̶ ̶a̶p̶p̶e̶a̶r̶ ̶i̶n̶ ̶t̶h̶e̶ ̶D̶a̶t̶a̶d̶o̶g̶ ̶U̶I̶ ̶u̶n̶t̶i̶l̶ ̶t̶h̶e̶ ̶n̶e̶x̶t̶ ̶d̶a̶t̶a̶d̶o̶g̶ ̶a̶g̶e̶n̶t̶ ̶r̶e̶l̶e̶a̶s̶e̶.̶ ̶

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- [x] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

File: ddtrace/contrib/starlette/patch.py [hunk 1]
```diff
@@ -129,3 +140,24 @@ def traced_handler(wrapped, instance, args, kwargs):
         )
 
     return wrapped(*args, **kwargs)
+
+
+@with_traced_module
+def _trace_background_tasks(module, pin, wrapped, instance, args, kwargs):
+    pin = Pin.get_from(module)
+    if not pin or not pin.enabled():
+        return wrapped(*args, **kwargs)
```
Comment (erikayasuda) [inline ddtrace/contrib/starlette/patch.py:169]: Can be removed (already checked in both `FastAPI` and `Starlette`)

File: ddtrace/contrib/starlette/patch.py [hunk 2]
```diff
@@ -150,3 +160,24 @@ def traced_handler(wrapped, instance, args, kwargs):
...
+
```
Comment (mabdinur) [inline ddtrace/contrib/starlette/patch.py:170]:
```suggestion
```

File: ddtrace/contrib/starlette/patch.py (hunk 1 above)
Comment (mabdinur) [reply ddtrace/contrib/starlette/patch.py]: Good catch 🤝

File: tests/contrib/fastapi/test_fastapi.py [hunk 3]
```diff
@@ -606,13 +620,15 @@ def test_table_query_snapshot(snapshot_client):
     }
 
 
-def test_background_task(client, tracer, test_spans):
-    """Tests if background tasks have been excluded from span duration"""
-    response = client.get("/asynctask")
+@snapshot(ignores=["meta._dd.span_links"])
```
Comment (erikayasuda) [inline tests/contrib/fastapi/test_fastapi.py:623]: Ignoring the span link attributes for now, because these are not [normalized like the span ID and trace IDs in traces](https://github.com/DataDog/dd-apm-test-agent/blob/97d78ff8478737f164982b58b867737b33fbc69b/ddapm_test_agent/trace_snapshot.py#L86) by the test agent. Issue is tracked [here](https://github.com/DataDog/dd-apm-test-agent/issues/154)

File: ddtrace/contrib/starlette/patch.py [hunk 4]
```diff
@@ -150,3 +160,20 @@ def traced_handler(wrapped, instance, args, kwargs):
...
+    task = get_argument_value(args, kwargs, 0, "func")
+    current_span = pin.tracer.current_span()
+
+    async def traced_task(*args, **kwargs):
+        with pin.tracer.start_span(
+            f"{module.__name__}.background_task", resource=task.__name__, child_of=None, activate=True
+        ) as span:
+            if current_span:
+                span.link_span(current_span.context)
+            await task(*args, **kwargs)
```
Comment (mabdinur) [inline ddtrace/contrib/starlette/patch.py:176]: The task isn't always await-able. Fastapi checks if this function is async and we should do the same: https://github.com/tiangolo/starlette/blob/654a65b6495862bbbffaaad8e9b04c8d50653451/starlette/background.py#L18

```suggestion
           if inspect. isawaitable(task)
               await task(*args, **kwargs)
           else:
               .... # make function awaitable? 
               
```

Comment (mabdinur) [reply ddtrace/contrib/starlette/patch.py:176]: This should address this test failure: https://github.com/DataDog/dd-trace-py/actions/runs/7645610961/job/20832607414?pr=7618

Comment (erikayasuda) [reply ddtrace/contrib/starlette/patch.py]: OK I'll try some things with what you have started here 👍

File: ddtrace/contrib/fastapi/patch.py [hunk 5]
```diff
@@ -52,7 +53,7 @@ async def traced_serialize_response(wrapped, instance, args, kwargs):
     the result.
     """
     pin = Pin.get_from(fastapi)
-    if not pin or not pin.enabled:
+    if not pin or not pin.enabled():
```
Comment (emmettbutler) [inline ddtrace/contrib/fastapi/patch.py:56]: Crazy that this little bug has been sitting in `main` for a while!

File: ddtrace/contrib/starlette/patch.py [hunk 6]
```diff
@@ -55,13 +60,17 @@ def patch():
     starlette._datadog_patch = True
 
     _w("starlette.applications", "Starlette.__init__", traced_init)
+    Pin().onto(starlette)
```
Comment (emmettbutler) [inline ddtrace/contrib/starlette/patch.py:63]: Why is this necessary?

**Review by emmettbutler — APPROVED** [review]
Great work. I left a few non-blocking comments.

File: ddtrace/contrib/starlette/patch.py (hunk 6 above)
Comment (erikayasuda) [reply ddtrace/contrib/starlette/patch.py:63]: @mabdinur can correct me if I'm wrong, but my understanding was that we need this for the `_trace_background_tasks()` function when we're trying to create the span links between the original task and the background task. We pass in the entire `starlette`/`fastapi` module, so we needed the pin to be associated with them at the module level.

File: ddtrace/contrib/fastapi/patch.py (hunk 5 above)
Comment (erikayasuda) [reply ddtrace/contrib/fastapi/patch.py:56]: Credit goes to @mabdinur for finding this 😄

File: ddtrace/contrib/starlette/patch.py (hunk 6 above)
Comment (mabdinur) [reply ddtrace/contrib/starlette/patch.py:63]: Yup. It's a bug. The starlete integration is not creating a pin but it is referencing it in different places. 

We could've pull this one line out into its own PR but I got lazy
